lib/Frontend/ArgsToFrontendOptionsConverter.cpp

```diff
@@ -413,145 +416,33 @@ bool ArgsToFrontendOptionsConverter::computeFallbackModuleName() {
+          ? outputFilenames->front()
           : Opts.InputsAndOutputs.getFilenameOfFirstInput().str();
+
   Opts.ModuleName = llvm::sys::path::stem(nameToStem);
   return false;
 }
 
 bool ArgsToFrontendOptionsConverter::computeOutputFilenames() {
-  assert(Opts.OutputFilenames.empty() &&
-         "Output filename should not be set at this point");
-  if (!FrontendOptions::doesActionProduceOutput(Opts.RequestedAction)) {
-    return false;
-  }
-  ArrayRef<std::string> outputFilenamesFromCommandLineOrFilelist =
-      getOutputFilenamesFromCommandLineOrFilelist();
-
-  if (outputFilenamesFromCommandLineOrFilelist.size() > 1) {
-    // WMO, threaded with N files (also someday batch mode).
-    Opts.OutputFilenames = outputFilenamesFromCommandLineOrFilelist;
-    return false;
-  }
-
-  if (outputFilenamesFromCommandLineOrFilelist.empty()) {
-    // When the Frontend is invoked without going through the driver
-    // (e.g. for testing), it is convenient to derive output filenames from
-    // input.
-    return deriveOutputFilenameFromInputFile();
-  }
-
-  StringRef outputFilename = outputFilenamesFromCommandLineOrFilelist[0];
-  if (!llvm::sys::fs::is_directory(outputFilename)) {
-    // Could be -primary-file (1), or -wmo (non-threaded w/ N (input) files)
-    Opts.OutputFilenames = outputFilenamesFromCommandLineOrFilelist;
-    return false;
-  }
-  // Only used for testing & when invoking frontend directly.
-  return deriveOutputFilenameForDirectory(outputFilename);
-}
-
-bool ArgsToFrontendOptionsConverter::deriveOutputFilenameFromInputFile() {
-  if (Opts.InputsAndOutputs.isReadingFromStdin() ||
-      FrontendOptions::doesActionProduceTextualOutput(Opts.RequestedAction)) {
-    Opts.setOutputFilenameToStdout();
-    return false;
-  }
-  std::string baseName = determineBaseNameOfOutput();
-  if (baseName.empty()) {
-    if (Opts.RequestedAction != FrontendOptions::ActionType::REPL &&
-        Opts.RequestedAction != FrontendOptions::ActionType::Immediate &&
-        Opts.RequestedAction != FrontendOptions::ActionType::NoneAction) {
-      Diags.diagnose(SourceLoc(), diag::error_no_output_filename_specified);
-      return true;
-    }
-    return false;
-  }
-  deriveOutputFilenameFromParts("", baseName);
-  return false;
-}
-
-bool ArgsToFrontendOptionsConverter::deriveOutputFilenameForDirectory(
-    StringRef outputDir) {
-
-  std::string baseName = determineBaseNameOfOutput();
-  if (baseName.empty()) {
-    Diags.diagnose(SourceLoc(), diag::error_implicit_output_file_is_directory,
-                   outputDir);
+  Optional<std::vector<std::string>> outs =
+      ArgsToFrontendOutputsConverter(Args, Opts.ModuleName,
+                                     Opts.InputsAndOutputs, Diags)
+          .convert();
+  if (!outs)
     return true;
-  }
-  deriveOutputFilenameFromParts(outputDir, baseName);
+  if (FrontendOptions::doesActionProduceOutput(Opts.RequestedAction))
+    Opts.InputsAndOutputs.setMainOutputs(*outs);
```

Comment:
Not sure I follow. Since the compiler supports NoneAction, Immediate, and REPL, there can be work done for no 'output'. Did you mean to suggest changing `doesActionProduceOutput` to `doesActionProduceOutputFiles`?

Reply:
No, I just mean that the (implicit) `else` branch of this throws out `outs`. Is that correct? Would it make sense to do this check first?

Reply:
I think it is correct to throw out outs in the else branch because there are not supposed to be any outputs. (The `ArgsToFrontendOutputsConverter` should issue a diagnostic if there are.)
If the `doesActionProduceOutput` check occurs first and the function returns if the answer is `false`, then `ArgsToFrontendOutputsConverter` won't get a chance to produce diagnostics.  Or perhaps I misunderstand?

Reply:
That is non-obvious from reading the code. Maybe in that case we know `outs` will be non-None, so we can test that instead?

Reply:
When you wrote "that is non-obvious" I'm not sure which of my statements you meant. Was it
"correct to throw out outs in the else branch because there are not supposed to be any outputs" or was it "then ArgsToFrontendOutputsConverter won't get a chance to produce diagnostics." or was it both? I know the code too well to see how it reads to others, sigh.

Reply:
I think we fixed it today.